File: bindings/dotnet/src/Wcl/Eval/WclValue.cs
using System;
using System.Collections.Generic;
using System.Text;
using Wcl.Core;

namespace Wcl.Eval
{
    public enum WclValueKind
    {
        String, Int, Float, Bool, Null,
        List, Map, Set, BlockRef
    }

    public class WclValue : IEquatable<WclValue>
    {
        public WclValueKind Kind { get; }

        private readonly string? _stringValue;
        private readonly long _intValue;
        private readonly double _floatValue;
        private readonly bool _boolValue;
        private readonly List<WclValue>? _listValue;
        private readonly OrderedMap<string, WclValue>? _mapValue;
        private readonly BlockRef? _blockRef;

        private WclValue(WclValueKind kind, string? s = null, long i = 0, double d = 0,
                         bool b = false, List<WclValue>? list = null,
                         OrderedMap<string, WclValue>? map = null,
                         BlockRef? br = null)
        {
            Kind = kind;
            _stringValue = s;
            _intValue = i;
            _floatValue = d;
            _boolValue = b;
            _listValue = list;
            _mapValue = map;
            _blockRef = br;
        }

        // Factory methods
        public static WclValue NewString(string value) => new WclValue(WclValueKind.String, s: value);
        public static WclValue NewInt(long value) => new WclValue(WclValueKind.Int, i: value);
        public static WclValue NewFloat(double value) => new WclValue(WclValueKind.Float, d: value);
        public static WclValue NewBool(bool value) => new WclValue(WclValueKind.Bool, b: value);
        public static readonly WclValue Null = new WclValue(WclValueKind.Null);
        public static WclValue NewList(List<WclValue> items) => new WclValue(WclValueKind.List, list: items);
        public static WclValue NewMap(OrderedMap<string, WclValue> map) => new WclValue(WclValueKind.Map, map: map);
        public static WclValue NewSet(List<WclValue> items) => new WclValue(WclValueKind.Set, list: items);
        public static WclValue NewBlockRef(BlockRef blockRef) => new WclValue(WclValueKind.BlockRef, br: blockRef);

        // Accessors
        public string AsString() => Kind == WclValueKind.String ? _stringValue! : throw new InvalidOperationException($"expected string, got {TypeName}");
        public long AsInt() => Kind == WclValueKind.Int ? _intValue : throw new InvalidOperationException($"expected int, got {TypeName}");
        public double AsFloat() => Kind == WclValueKind.Float ? _floatValue : throw new InvalidOperationException($"expected float, got {TypeName}");
        public bool AsBool() => Kind == WclValueKind.Bool ? _boolValue : throw new InvalidOperationException($"expected bool, got {TypeName}");
        public List<WclValue> AsList() => Kind == WclValueKind.List ? _listValue! : throw new InvalidOperationException($"expected list, got {TypeName}");
        public OrderedMap<string, WclValue> AsMap() => Kind == WclValueKind.Map ? _mapValue! : throw new InvalidOperationException($"expected map, got {TypeName}");
        public List<WclValue> AsSet() => Kind == WclValueKind.Set ? _listValue! : throw new InvalidOperationException($"expected set, got {TypeName}");
        public BlockRef AsBlockRef() => Kind == WclValueKind.BlockRef ? _blockRef! : throw new InvalidOperationException($"expected block_ref, got {TypeName}");

        // Try accessors
        public string? TryAsString() => Kind == WclValueKind.String ? _stringValue : null;
        public long? TryAsInt() => Kind == WclValueKind.Int ? _intValue : (long?)null;
        public double? TryAsFloat() => Kind == WclValueKind.Float ? _floatValue : (double?)null;
        public bool? TryAsBool() => Kind == WclValueKind.Bool ? _boolValue : (bool?)null;
        public List<WclValue>? TryAsList() => Kind == WclValueKind.List ? _listValue : null;
        public OrderedMap<string, WclValue>? TryAsMap() => Kind == WclValueKind.Map ? _mapValue : null;
        public BlockRef? TryAsBlockRef() => Kind == WclValueKind.BlockRef ? _blockRef : null;

        public bool IsNull => Kind == WclValueKind.Null;

        public bool? IsTruthy() => Kind == WclValueKind.Bool ? _boolValue : (bool?)null;

        public string TypeName => Kind switch
        {
            WclValueKind.String => "string",
            WclValueKind.Int => "int",
            WclValueKind.Float => "float",
            WclValueKind.Bool => "bool",
            WclValueKind.Null => "null",
            WclValueKind.List => "list",
            WclValueKind.Map => "map",
            WclValueKind.Set => "set",
            WclValueKind.BlockRef => "block_ref",
            _ => "unknown",
        };

        public string ToInterpString()
        {
            switch (Kind)
            {
                case WclValueKind.String: return _stringValue!;
                case WclValueKind.Int: return _intValue.ToString();
                case WclValueKind.Float: return _floatValue.ToString(System.Globalization.CultureInfo.InvariantCulture);
                case WclValueKind.Bool: return _boolValue ? "true" : "false";
                case WclValueKind.Null: return "null";
                default: throw new InvalidOperationException($"cannot interpolate {TypeName} into string");
            }
        }

        public bool Equals(WclValue? other)
        {
            if (other is null) return false;
            if (Kind != other.Kind) return false;
            switch (Kind)
            {
                case WclValueKind.String: return _stringValue == other._stringValue;
                case WclValueKind.Int: return _intValue == other._intValue;
                case WclValueKind.Float: return _floatValue == other._floatValue;
                case WclValueKind.Bool: return _boolValue == other._boolValue;
                case WclValueKind.Null: return true;
                case WclValueKind.List:
                case WclValueKind.Set:
                {
                    if (_listValue!.Count != other._listValue!.Count) return false;
                    for (int i = 0; i < _listValue.Count; i++)
                        if (!_listValue[i].Equals(other._listValue[i])) return false;
                    return true;
                }
                case WclValueKind.Map:
                {
                    if (_mapValue!.Count != other._mapValue!.Count) return false;
                    for (int i = 0; i < _mapValue.Count; i++)
                    {
                        var a = _mapValue.GetAt(i);
                        var b = other._mapValue.GetAt(i);
                        if (a.Key != b.Key || !a.Value.Equals(b.Value)) return false;
                    }
                    return true;
                }
                default: return false;
            }
        }

        public override bool Equals(object? obj) => obj is WclValue other && Equals(other);
        public override int GetHashCode() => Kind.GetHashCode();

        public static bool operator ==(WclValue? left, WclValue? right)
        {
            if (left is null) return right is null;
            return left.Equals(right);
        }
        public static bool operator !=(WclValue? left, WclValue? right) => !(left == right);

        public override string ToString()
        {
            switch (Kind)
            {
                case WclValueKind.String: return _stringValue!;
                case WclValueKind.Int: return _intValue.ToString();
                case WclValueKind.Float: return _floatValue.ToString(System.Globalization.CultureInfo.InvariantCulture);
                case WclValueKind.Bool: return _boolValue ? "true" : "false";
                case WclValueKind.Null: return "null";
                case WclValueKind.List:
                {
                    var sb = new StringBuilder("[");
                    for (int i = 0; i < _listValue!.Count; i++)
                    {
                        if (i > 0) sb.Append(", ");
                        sb.Append(_listValue[i]);
                    }
                    sb.Append(']');
                    return sb.ToString();
                }
                case WclValueKind.Map:
                {
                    var sb = new StringBuilder("{");
                    int idx = 0;
                    foreach (var kvp in _mapValue!)
                    {
                        if (idx++ > 0) sb.Append(", ");
                        sb.Append(kvp.Key).Append(" = ").Append(kvp.Value);
                    }
                    sb.Append('}');
                    return sb.ToString();
                }
                case WclValueKind.Set:
                {
                    var sb = new StringBuilder("set(");
                    for (int i = 0; i < _listValue!.Count; i++)
                    {
                        if (i > 0) sb.Append(", ");
                        sb.Append(_listValue[i]);
                    }
                    sb.Append(')');
                    return sb.ToString();
                }
                case WclValueKind.BlockRef:
                {
                    var br = _blockRef!;
                    var sb = new StringBuilder(br.Kind);
                    if (br.Id != null) sb.Append(' ').Append(br.Id);
                    sb.Append(" {");
                    int idx = 0;
                    foreach (var kvp in br.Attributes)
                    {
                        if (idx++ > 0) sb.Append(',');
                        sb.Append(' ').Append(kvp.Key).Append(" = ").Append(kvp.Value);
                    }
                    sb.Append(" }");
                    return sb.ToString();
                }
                default: return Kind.ToString();
            }
        }
    }
}

File: bindings/jvm/src/main/java/io/github/wiltaylor/wcl/eval/WclValue.java
package io.github.wiltaylor.wcl.eval;

import java.util.*;

public final class WclValue {
    private final WclValueKind kind;
    private final String stringValue;
    private final long intValue;
    private final double floatValue;
    private final boolean boolValue;
    private final List<WclValue> listValue;
    private final LinkedHashMap<String, WclValue> mapValue;
    private final BlockRef blockRef;

    public static final WclValue NULL = new WclValue(WclValueKind.NULL, null, 0, 0, false, null, null, null);

    private WclValue(WclValueKind kind, String s, long i, double d, boolean b,
                     List<WclValue> list, LinkedHashMap<String, WclValue> map, BlockRef br) {
        this.kind = kind;
        this.stringValue = s;
        this.intValue = i;
        this.floatValue = d;
        this.boolValue = b;
        this.listValue = list;
        this.mapValue = map;
        this.blockRef = br;
    }

    // Factory methods
    public static WclValue ofString(String value) {
        return new WclValue(WclValueKind.STRING, value, 0, 0, false, null, null, null);
    }

    public static WclValue ofInt(long value) {
        return new WclValue(WclValueKind.INT, null, value, 0, false, null, null, null);
    }

    public static WclValue ofFloat(double value) {
        return new WclValue(WclValueKind.FLOAT, null, 0, value, false, null, null, null);
    }

    public static WclValue ofBool(boolean value) {
        return new WclValue(WclValueKind.BOOL, null, 0, 0, value, null, null, null);
    }

    public static WclValue ofList(List<WclValue> items) {
        return new WclValue(WclValueKind.LIST, null, 0, 0, false, items, null, null);
    }

    public static WclValue ofMap(LinkedHashMap<String, WclValue> map) {
        return new WclValue(WclValueKind.MAP, null, 0, 0, false, null, map, null);
    }

    public static WclValue ofSet(List<WclValue> items) {
        return new WclValue(WclValueKind.SET, null, 0, 0, false, items, null, null);
    }

    public static WclValue ofBlockRef(BlockRef blockRef) {
        return new WclValue(WclValueKind.BLOCK_REF, null, 0, 0, false, null, null, blockRef);
    }

    public static WclValue ofBigInt(long value) {
        return new WclValue(WclValueKind.BIG_INT, null, value, 0, false, null, null, null);
    }

    public static WclValue ofDate(String value) {
        return new WclValue(WclValueKind.DATE, value, 0, 0, false, null, null, null);
    }

    public static WclValue ofDuration(String value) {
        return new WclValue(WclValueKind.DURATION, value, 0, 0, false, null, null, null);
    }

    // Accessors
    public WclValueKind getKind() { return kind; }

    public String asString() {
        if (kind != WclValueKind.STRING) throw new IllegalStateException("expected string, got " + typeName());
        return stringValue;
    }

    public long asInt() {
        if (kind != WclValueKind.INT) throw new IllegalStateException("expected int, got " + typeName());
        return intValue;
    }

    public double asFloat() {
        if (kind != WclValueKind.FLOAT) throw new IllegalStateException("expected float, got " + typeName());
        return floatValue;
    }

    public boolean asBool() {
        if (kind != WclValueKind.BOOL) throw new IllegalStateException("expected bool, got " + typeName());
        return boolValue;
    }

    public List<WclValue> asList() {
        if (kind != WclValueKind.LIST) throw new IllegalStateException("expected list, got " + typeName());
        return listValue;
    }

    public LinkedHashMap<String, WclValue> asMap() {
        if (kind != WclValueKind.MAP) throw new IllegalStateException("expected map, got " + typeName());
        return mapValue;
    }

    public List<WclValue> asSet() {
        if (kind != WclValueKind.SET) throw new IllegalStateException("expected set, got " + typeName());
        return listValue;
    }

    public BlockRef asBlockRef() {
        if (kind != WclValueKind.BLOCK_REF) throw new IllegalStateException("expected block_ref, got " + typeName());
        return blockRef;
    }

    public long asBigInt() {
        if (kind != WclValueKind.BIG_INT) throw new IllegalStateException("expected bigint, got " + typeName());
        return intValue;
    }

    public String asDate() {
        if (kind != WclValueKind.DATE) throw new IllegalStateException("expected date, got " + typeName());
        return stringValue;
    }

    public String asDuration() {
        if (kind != WclValueKind.DURATION) throw new IllegalStateException("expected duration, got " + typeName());
        return stringValue;
    }

    // Try accessors
    public Optional<String> tryAsString() {
        return kind == WclValueKind.STRING ? Optional.of(stringValue) : Optional.empty();
    }

    public OptionalLong tryAsInt() {
        return kind == WclValueKind.INT ? OptionalLong.of(intValue) : OptionalLong.empty();
    }

    public OptionalDouble tryAsFloat() {
        return kind == WclValueKind.FLOAT ? OptionalDouble.of(floatValue) : OptionalDouble.empty();
    }

    public boolean isNull() {
        return kind == WclValueKind.NULL;
    }

    public String typeName() {
        return switch (kind) {
            case STRING -> "string";
            case INT -> "int";
            case FLOAT -> "float";
            case BOOL -> "bool";
            case NULL -> "null";
            case LIST -> "list";
            case MAP -> "map";
            case SET -> "set";
            case BLOCK_REF -> "block_ref";
            case BIG_INT -> "bigint";
            case DATE -> "date";
            case DURATION -> "duration";
        };
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof WclValue other)) return false;
        if (kind != other.kind) return false;
        return switch (kind) {
            case STRING, DATE, DURATION -> Objects.equals(stringValue, other.stringValue);
            case INT, BIG_INT -> intValue == other.intValue;
            case FLOAT -> Double.compare(floatValue, other.floatValue) == 0;
            case BOOL -> boolValue == other.boolValue;
            case NULL -> true;
            case LIST, SET -> Objects.equals(listValue, other.listValue);
            case MAP -> Objects.equals(mapValue, other.mapValue);
            case BLOCK_REF -> Objects.equals(blockRef, other.blockRef);
        };
    }

    @Override
    public int hashCode() {
        return kind.hashCode();
    }

    @Override
    public String toString() {
        return switch (kind) {
            case STRING, DATE, DURATION -> stringValue;
            case INT, BIG_INT -> Long.toString(intValue);
            case FLOAT -> Double.toString(floatValue);
            case BOOL -> boolValue ? "true" : "false";
            case NULL -> "null";
            case LIST -> {
                var sb = new StringBuilder("[");
                for (int i = 0; i < listValue.size(); i++) {
                    if (i > 0) sb.append(", ");
                    sb.append(listValue.get(i));
                }
                sb.append(']');
                yield sb.toString();
            }
            case MAP -> {
                var sb = new StringBuilder("{");
                int idx = 0;
                for (var entry : mapValue.entrySet()) {
                    if (idx++ > 0) sb.append(", ");
                    sb.append(entry.getKey()).append(" = ").append(entry.getValue());
                }
                sb.append('}');
                yield sb.toString();
            }
            case SET -> {
                var sb = new StringBuilder("set(");
                for (int i = 0; i < listValue.size(); i++) {
                    if (i > 0) sb.append(", ");
                    sb.append(listValue.get(i));
                }
                sb.append(')');
                yield sb.toString();
            }
            case BLOCK_REF -> {
                var sb = new StringBuilder(blockRef.getKind());
                if (blockRef.getId() != null) sb.append(' ').append(blockRef.getId());
                sb.append(" {");
                int idx = 0;
                for (var entry : blockRef.getAttributes().entrySet()) {
                    if (idx++ > 0) sb.append(',');
                    sb.append(' ').append(entry.getKey()).append(" = ").append(entry.getValue());
                }
                sb.append(" }");
                yield sb.toString();
            }
        };
    }
}
